### ISSUE TITLE
Add factory for exceptions 

### ISSUE THREAD

**Issue by bluemmc** [swagger-api/swagger-codegen#2724]
I am using a rest api which returns one type of JSON for status 200 (ok) and another JSON for status>=400 (errors). The JSON returned for HTTP errors give detailed error information.

Currently, the generated API (at least for C#) is hardcoded to throw ApiExpceptions for status>=400 with the json returned as text. It would be nice if I can override this behavior to return a typesafe exception with a deserialized json object. Such behavour is client-specific, but could be supported if the generated api can be configured with a factory method that is responsible for throwing the exception. Then a client like mine can override the default behavior.

**Comment by jimschubert** [comment]
Please correct me if I'm wrong, but I'll be speaking to this code in *Api.cs classes:
```
if (localVarStatusCode >= 400)
    throw new ApiException (localVarStatusCode, "Error calling myOperation: " + localVarResponse.Content, localVarResponse.Content);
else if (localVarStatusCode == 0)
    throw new ApiException (localVarStatusCode, "Error calling myOperation: " + localVarResponse.ErrorMessage, localVarResponse.ErrorMessage);
```

Are you saying you want a factory method on each *Api type (implemented as a partial method), a specific factory method for each operation (could be implemented as a partial method, one per operation), or a global factory method (on the `Configuration` object)?

I agree with the suggestion that we could use a little more control here. I like to design JSON-based APIs with a **problem result** as defined by https://tools.ietf.org/html/draft-nottingham-http-problem-06. An example response:

```
   HTTP/1.1 403 Forbidden
   Content-Type: application/problem+json
   Content-Language: en

   {
    "type": "http://example.com/probs/out-of-credit",
    "title": "You do not have enough credit.",
    "detail": "Your current balance is 30, but that costs 50.",
    "instance": "http://example.net/account/12345/msgs/abc",
    "balance": 30,
    "accounts": ["http://example.net/account/12345",
                 "http://example.net/account/67890"]
   }
```

It would be a lot nicer to have the ability to do something on a per-Api basis:

```
// …
if (localVarStatusCode >= 400)
    throw CustomError("myOperation", response);
// …
```

The problem is that this leaves you with modifying generated classes. I'm in the process of making Api classes partials, which opens us up for adding partial methods. A partial method, however, makes the above construct a little clunky:

```
partial void InspectResponse(IRestResponse response); // May throw exception

// …
InspectResponse(response);
if (localVarStatusCode >= 400)
    throw new ApiException(localVarResponse)
// …
```
This would require you to implement `InspectResponse` on every type with a custom exception.

An alternative would be to add some error handler construct to `Configuration`, which can be passed to the Api type. This would make it a little cleaner in the Api classes:

```
// …
if (localVarStatusCode >= 400)
    throw Configuration.BuildResponseException(localVarResponse);
// …
```
This would work and is probably the easiest to work with, but I think it could end up being used in unintended ways. You could also end up with lots of custom exception factories if the API endpoints all return different objects for different errors.

**Comment by bluemmc** [comment]
I suggest the following (untested but I think it will work):

1) Declare this type globally:
`
public delegate Exception ExceptionFactory(int localVarStatusCode, IRestResponse response);
`
2) Create an instance level property on the configuration (here called ExceptionFactory)
As a default the property should be set to this value which is backward compatible with your current approach:
`
 ExceptionFactory= (_methodName, _localVarStatusCode, _response) =>
            {
                if (_localVarStatusCode >= 400)
                    return new ApiException(_localVarStatusCode, "Error calling "+_methodName+": " + _response.Content, _response.Content);
                else if (_localVarStatusCode == 0)
                    return new ApiException(_localVarStatusCode, "Error calling "+_methodName+": " + _response.ErrorMessage, _response.ErrorMessage);
                else return null; // Ignore all other codes.
            };
`

Then change this presently generated code
`
if (localVarStatusCode >= 400)
    throw new ApiException (localVarStatusCode, "Error calling myOperation: " + localVarResponse.Content, localVarResponse.Content);
else if (localVarStatusCode == 0)
    throw new ApiException (localVarStatusCode, "Error calling myOperation: " + localVarResponse.ErrorMessage, localVarResponse.ErrorMessage);
`

To code like this:
`
if (localVarStatusCode >= 400 || localVarStatusCode == 0)
{
  var exception = Configuration.ExceptionFactory("myOperation", localVarStatusCode, localVarResponse);
  if (exception != null)
  throw exception;
}
`

With the above approach, the exceptions for an endpoint can be customized or suppressed by changing the configuration. With your other interface change in  #2723 it is trivial to create general code that does the same for a number of endpoints if a global change is wanted.

**Comment by bluemmc** [comment]
BTW: We could also consider removing the last if-statement in the generated code as the suggested factory checks for error.

**Comment by jimschubert** [comment]
@bluemmc I implemented the ExceptionFactory by merging both of our ideas. I couldn't find a clean way to prevent a multicast delegate, so I just check the invocation list of the delegate and throw an exception. I also removed the status code parameter because it's already a part of the response.

Please check out the branch and tell me what you think!

**Comment by bluemmc** [comment]
Great. Minor points for the branch:
1) As I understand it the InterceptResponse bits are fully independent from the ExceptionFactory  bits. Are they for two different use-cases?
2) Maybe move the assignment to localVarStatusCode below the call to ExceptionFactory in api.mustache (in case the ExceptionFactory modifies the response to cope with the error - not sure this practice of modifying the response is a good idea in the factory but if it is done then moving the assignment will make it work).

**Comment by bluemmc** [comment]
BTW: Let me know when the feature has been merged with the main branch. Then I will try it out (for now I have only inspected the source code of the change).

**Comment by wing328** [comment]
@bluemmc To try the branch locally, please run the following:
```
git checkout -b jimschubert-cs/exception_factory master
git pull https://github.com/jimschubert/swagger-codegen.git cs/exception_factory
mvn clean package
```

**Comment by jimschubert** [comment]
@bluemmc 

Yes, `InterceptRequest`, `InterceptResponse` and `ExceptionFactory` are all different use cases. The `InterceptRequest/InterceptResponse` methods are partial methods. This means you'd need to implement a function which gets compiled as part of the `ApiClient` type definition.  Taking the example in the pull request and modifying it a bit to use Trace instead of Debug:

```
// in src/IO.Swagger/Client/ApiClientRequestInterceptor.cs
//…
public partial class ApiClient 
{
    partial void InterceptRequest(IRestRequest request)
    {
        Trace.WriteLine("Request made: {0}", request.Path);
    }
}
```

This would allow consumers to apply a trace listener to `ApiClient` and get this diagnostic information for every request. If you use zipkin in your system, you could also implement both methods and add the required headers to the request then use those headers to trace statistics about round trip.

Likewise, you could use the `InterceptRequest` method to guarantee required headers are in the correct format.

The `ExceptionFactory` on the other hand would act on an ApiClient instance level. You can use the client library with a single global instance of `ApiClient`, but you can also construct a `Configuration` instance and assign it a new `ApiClient` then pass that into the constructor of the API classes.

This would look something like this:

```
var apiClient = new ApiClient("http://petstore.swagger.io/v2");

// 
apiClient.ExceptionFactory = (methodName, response) => {
    int status = (int) response.StatusCode;
    // NOTE: Throwing here is basically the same as returning the exception to the calling method.
    // Exceptions are specific to the consumer's API, not generic as in the default
    if(status == 400) throw new MyOtherFrameworkBadRequestException(response.body);
    if(status == 404) throw new MyOtherFrameworkNotFoundException(response.body);

    // Allows code to handle missing data (untested, used as an example)
    if((int)(status/100) == 5) {
        response.Content = null;
        return null;
    }
};

var config = new Configuration(apiClient);
var petApi = new PetApi(config);
var pets = petApi.FindPetsByStatus(new List<String> { "active" });
if(pets == null || pets.Length > 0) {
    // process pets
} else {
    // we don't care if pets weren't found or if there was a 5xx error
}
```

Obviously in the above example you'd want to use the InterceptResponse method to do some logging for better insight into issues. With microservices architectures becoming more and more common, a single request/response flow that fails doesn't indicate a failure… you could implement retries with incremental backoff (this is why `InterceptResponse` takes both the request and response instances, it's clearer even though response.Request exists). A 5xx response could be a result of some dependency downstream being unavailable, often systems are written to account for this and a client api library rigidly throwing exceptions in these cases isn't always the best solution.

Also for many APIs, data is not a necessity. For example, in this issue we all have avatars or identicons. The identicons (like @bluemmc's) is generated against some account information like username, and is served from github. My avatar is synced from gravatar. Suppose the query for my avatar comes from a different system than the query for your identicon. If that system goes down and throws a 5xx error of any kind, you have two options: fail the entire request or gracefully handle the missing image. The example above would allow us to handle the missing image. Suppose we wanted to fail the entire request. The default `ExceptionFactory` would throw a type of `Your.Package.Client.ApiException` which wouldn't be the same exception as if the identicon server accessed via some other client library responded with a 5xx error. You'd end up catching and handling multiple exceptions for nicety data. But what if the gravatar is the only endpoint you don't care about a 5xx for?

```
apiClient.ExceptionFactory = (methodName, response) => {
    int status = (int) response.StatusCode;
    // Handle all 5xx messages
    if((int)(status/100) == 5 && response.Request.Resource == "/gravatar/{id}") {
        response.Content = null;
        return null;
    }
    
    // defer to default factory
    return Configuration.DefaultExceptionFactory(methodName, response);
};
```

Let me know if that doesn't fully answer the question, or if any of that doesn't make sense.

**Comment by bluemmc** [comment]
@jimschubert 
Thanks for the detailed explanation.Sounds good. The only potential drawback I can think of is that the use of partical classes to override InterceptRequest is c# specific and thus difficult to align with other languages if they want to implement support for intercepters too. Apart from that I think your contribution with intercepters and the exception factory is great and I look forward to see it in the main branch and in the next release.

**Comment by bluemmc** [comment]
@jimschubert 
What are the plans for merging this with main branch?

**Comment by wing328** [comment]
@bluemmc let me review the PR today or tomorrow. (have been really busy recently ...)

**Comment by jackjwilliams** [comment]
I realize this is closed and merged, but I recently tried to use InterceptRequest in a client project (that uses the codegen'd dll), but I could never get it to compile. Do I have to have a partial ApiClient class IN the Generated project or can I do it in my project that uses the dll? My usage of partial methods is limited, and I never could get this to work.

**Comment by jimschubert** [comment]
@jackjwilliams Yeah you need the partial class with the same name in your generated project. In C#, partial classes are all merged to make a single larger class. It's a simpler way to allow for code extensions than IoC, because we would have to support multiple IoC frameworks whereas partial classes/methods are a language feature.

**Comment by jackjwilliams** [comment]
Gotcha, so I would most likely want to create / override a template and make sure it's included in the generation. Thanks!
